src/Formidable.js

```diff
@@ -527,28 +536,45 @@ class IncomingForm extends EventEmitter {
     return basename.slice(firstDot, lastDot) + extname;
   }
 
-  _uploadPath(part, fp) {
-    const name = fp || `${this.uploadDir}${path.sep}${toHexoId()}`;
 
-    if (part && this.options.keepExtensions) {
-      const filename = typeof part === 'string' ? part : part.filename;
-      return `${name}${this._getExtension(filename)}`;
+
+  _joinDirectoryName(name) {
+    const newPath = path.join(this.uploadDir, name);
+
+    // prevent directory traversal attacks
+    if (!newPath.startsWith(this.uploadDir)) {
+      return path.join(this.uploadDir, this.options.defaultInvalidName);
     }
 
-    return name;
+    return newPath;
   }
 
   _setUpRename() {
     const hasRename = typeof this.options.filename === 'function';
-
-    if (this.options.keepExtensions === true && hasRename) {
-      this._rename = (part) => {
-        const resultFilepath = this.options.filename.call(this, part, this);
-
-        return this._uploadPath(part, resultFilepath);
+    if (hasRename) {
+      this._getNewName = (part) => {
+        let ext = '';
+        let name = this.options.defaultInvalidName;
+        if (part.originalFilename) {
+          // can be null
+          ({ ext, name } = path.parse(part.originalFilename));
+          if (this.options.keepExtensions !== true) {
+            ext = '';
+          }
+        }
+        return this.options.filename.call(this, name, ext, part, this);
```

Comment:
`path`'s return of `ext` isn't handling well filenames with multiple dots, that's why i did the `_getExtension` method, maybe use it here too.

```js
this._getNewName = (part) => {
  let filename = this.options.defaultInvalidName
  let extname = '';

  if (part.mimetype) {
    filename = part.originalFilename;
    if (this.options.keepExtensions === true) {
      extname = this._getExtename(part.originalFilename);
    }
  }

  return this.options.filename.call(this, filename, extname, part, this);
};
```

Reply:
but the problem is that there would be duplicated parts. For example x.y.z.w would call
this.options.filename.call(this, name, ext, part, this);
with name = 'x.y.z' and ext = '.y.z.w'